### PR TITLE
Add oidc-proxy backend auth provider

### PR DESCRIPTION
**Pull request by jeffmccune**
## Hey, I just made a Pull Request!

Add oidc-proxy backend auth provider (#25103)

Previously, the various proxy based auth providers are all tied to a specific identity aware proxy.  For example, the gcp-iap provider is tightly coupled to the GCP identity aware proxy.

This is a problem for similar architectures such as Holos Platform, which uses Istio with ExtAuthZ to provide an oidc id token in the `x-oidc-id-token` header.  The existing providers cannot be easily re-used for such an integration even though the only major difference is the OIDC issuer url and the header used to carry the id token.

This patch solves the problem by implementing a generally useful `auth-backend-module-oidc-proxy-provider` with a provider id of `'oidc-proxy'`.

This backend auth provider is implemented as a nearly perfect copy of the existing gcp-iap provider, which the following notable exceptions.

 1. Integrators may configure the oidc issuer with oidc-proxy using the `iss` Config field.
 2. Integrators may configure the header the id token is expected at using the `oidcIdTokenHeader` Config field.

Both of these settings are hard-coded to Google specific values through the use of the `google-auth-library` in the case of the `gcp-iap` provider.

Otherwise, this `oidc-proxy` provider strives to stay true to the spirit and behavior of the `gcp-iap` provider in an effort to reduce the cognitive load of the core maintainers and integrators.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Closes: #25103

### CODE REVIEW

**Comment by jeffmccune** [comment]
Marking this as draft until I add documentation and figure out how to test this properly as an integrate would following the recommended way to install a plugin.

To save me some time, please could you provide some guidance on how best to build a container image with this plugin?  In the meantime, I'll try and figure out it,

**Comment by jeffmccune** [comment]
@freben Please take a look when you have some time.  I'm really struggling to build and test this, at this point I've spent about twice as long wrestling with `yarn` than I spent writing the code and unit tests.

I have a fresh setup with `npx @backstage/create-app@latest` but I can't figure how how to get this provider linked in. 
 When adding a new core plugin like this, how should I build it into a container and test it before the PR gets merged?

**Comment by jeffmccune** [comment]
I was able to make some progress by taking the following approach:

```
npx @backstage/create-app@latest
? Enter a name for the app [required] portal
```

Then tried to create an internal backend module auth provider:

```
❯ yarn new --select backend-module
yarn run v1.22.19
$ backstage-cli new --scope internal --select backend-module
? Enter the ID of the plugin [required] auth
? Enter the ID of the module [required] holos-proxy-provider
```

I copied over all of the code here but used a different provider id of `holosProxy` instead of `oidcProxy`.  The change list is at https://github.com/holos-run/portal/pull/2/commits

But, the deployment still fails because I don't understand how to register the new auth provider like is done [here](https://github.com/backstage/backstage/pull/25105/files#diff-2c4c9015b1e5166def5c2014243e2bab9f9ac7745dd9fc4a094823f5bbb0001eR59)

<details><summary>No sign-in resolver configured for proxy auth provider 'holosProxy'</summary>

```
{"level":"info","message":"Configuring \"database\" as KeyStore provider","plugin":"auth","service":"backstage"}
{"level":"info","message":"Configuring auth provider: holosProxy","plugin":"auth","service":"backstage"}
/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1859
          throw new errors.ForwardedError(
                ^

ForwardedError: Plugin 'auth' startup failed; caused by Error: Failed to initialize holosProxy auth provider, No sign-in resolver configured for proxy auth provider 'holosProxy'
    at /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1859:17
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1858:9
    ... 3 lines matching cause stack trace ...
    at async BackstageBackend.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1949:5) {
  cause: Error: Failed to initialize holosProxy auth provider, No sign-in resolver configured for proxy auth provider 'holosProxy'
      at bindProviderRouters (/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:1722:17)
      at createRouter (/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:2468:3)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Object.init [as func] (/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:2535:24)
      at async /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1858:9
      at async Promise.all (index 4)
      at async BackendInitializer.doStart_fn (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1819:3)
      at async BackendInitializer.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1675:5)
      at async BackstageBackend.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1949:5)
}

Node.js v18.20.3
```
</details>

The container is built with:

```
# Build the backend
yarn install --frozen-lockfile
yarn tsc
yarn build:backend --config ../../app-config.yaml
# Build the image
docker image build . -f packages/backend/Dockerfile --tag quay.io/holos-run/portal:latest
docker push quay.io/holos-run/portal:latest
```

Only other gotcha which I think I got past was adding `node-fetch`:

```
yarn --cwd plugins/auth-backend-module-holos-proxy-provider add "node-fetch@^2.6.11"
yarn --cwd plugins/auth-backend-module-holos-proxy-provider add "@types/node-fetch@^2.6.11" --dev
```

**Comment by jeffmccune** [comment]
As an update, I've figured out how to deploy this and am able to work independently as of commit https://github.com/holos-run/portal/pull/2/commits/f5dfc92edd5a22d354a8312e0a25f626e0eebff6 in the https://github.com/holos-run/portal/pull/2/commits changeset.

When I browse to the deployment, I see the following request in the pod logs:

```json
{
  "level": "info",
  "message": "::ffff:127.0.0.6 - - [11/Jun/2024:02:24:25 +0000] \"GET /api/auth/holosProxy/refresh HTTP/1.1\" 401 224 \"-\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36\"",
  "service": "rootHttpRouter",
  "type": "incomingRequest"
}
```

The browser Network inspector indicates the problem I need to fix, likely with how this PR fetches the `jwks_uri` from the discovery document.

```json
{
    "error": {
        "name": "AuthenticationError",
        "message": "could not validate id token: Unsupported \"alg\" value for a JSON Web Key Set"
    },
    "request": {
        "method": "GET",
        "url": "/api/auth/holosProxy/refresh"
    },
    "response": {
        "statusCode": 401
    }
}
```

In any event, I think I'm on the right path now.

**Comment by jeffmccune** [comment]
Switching to draft for now.  The jwks loading has been fixed in https://github.com/holos-run/portal/pull/2/commits/923d908630de9109a85586521fcd34784e30de51 and I'm on a path where I can test and document this properly.

I'll re-open once the test plan and docs are ready for review.

**Comment by jeffmccune** [comment]
@freben  This `oidcProxy` we discussed is ready for review.  The method I used to test this is described in the commits in https://github.com/holos-run/portal/pull/2

Happy to make any changes necessary, thanks!

**Comment by awanlin** [comment]
Hi @jeffmccune, you're just hitting the Summer Vacation period so reviews may be a bit slow or sporadic. 👍

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface Config {
+  auth?: {
+    providers?: {
+      /**
+       * Configuration for the Holos Platform identity aware proxy auth provider.
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:21]: not that specific right?

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+       */
+      oidcProxy?: {
+        [authEnv: string]: {
+          /**
+           * issuer represents the iss claim value to assert validating an id token.
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:26]:
```suggestion
           * the iss claim value to assert validating an id token.
```

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+           */
+          issuer: string;
+
+          /**
+           * audience represents the aud claim value to assert when validating an id token.
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:31]:
```suggestion
           * the aud claim value to assert when validating an id token.
```

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+           */
+          audience: string;
+
+          /**
+           * idTokenHeader represents the header to read the id token from.  Defaults to `'x-oidc-id-token'`.
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:36]:
```suggestion
           * the header to read the id token from.  Defaults to `'x-oidc-id-token'`.
```

no need to repeat these; it'll just interfere with refactoring etc

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+export interface Config {
+  auth?: {
+    providers?: {
+      /**
+       * Configuration for the Holos Platform identity aware proxy auth provider.
+       */
+      oidcProxy?: {
+        [authEnv: string]: {
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:24]: This entire layer is usually not present on proxy providers. The `createProxyAuthenticator` function will try to locate config that does not have this extra nesting.

File: plugins/auth-backend-module-oidc-proxy-provider/package.json [hunk 6]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@backstage/plugin-auth-backend-module-oidc-proxy-provider",
+  "version": "0.1.0",
+  "description": "A Holos Platform auth provider module for the Backstage auth backend, generally useful with OIDC ID tokens set in HTTP request headers by external auth proxies.",
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/package.json:4]: same here, not really just for holos

File: plugins/auth-backend-module-oidc-proxy-provider/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Auth Backend Module - Holos Platform OIDC Proxy Provider
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/README.md:1]: less specific?

File: plugins/auth-backend-module-oidc-proxy-provider/CHANGELOG.md
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/CHANGELOG.md:1]: these files are auto generated by the release process; you can remove this

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import {
+  createProxyAuthenticator,
+  ProfileInfo,
+} from '@backstage/plugin-auth-node';
+import { createTokenValidator } from './helpers';
+import { OidcProxyResult } from './types';
+import { LoggerService } from '@backstage/backend-plugin-api';
+
+/**
+ * DEFAULT_OIDC_ID_TOKEN_HEADER represents the default http request header used
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:27]:
```suggestion
 * the default http request header used
```

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ * when an alternative header has not been explicitly configured.
+ */
+const DEFAULT_OIDC_ID_TOKEN_HEADER = 'x-oidc-id-token';
+
+export function createHolosProxyAuthenticator(logger: LoggerService) {
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:32]:
```suggestion
export function createOidcProxyAuthenticator(logger: LoggerService) {
```

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  return createProxyAuthenticator({
+    defaultProfileTransform: async (result: OidcProxyResult) => {
+      const profileInfo: ProfileInfo = { email: result.idToken.email };
+      if ('name' in result.idToken) {
+        profileInfo.displayName = result.idToken.name as string;
+      }
+      if ('picture' in result.idToken) {
+        profileInfo.picture = result.idToken.picture as string;
+      }
+      logger.debug(`holos profile transform: ${JSON.stringify(profileInfo)}`);
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:42]:
```suggestion
      logger.debug(`oidc proxy profile transform: ${JSON.stringify(profileInfo)}`);
```

File: plugins/auth-backend-module-oidc-proxy-provider/src/module.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          providerId: 'oidcProxy',
+          factory: createProxyAuthProviderFactory({
+            authenticator: createHolosProxyAuthenticator(logger),
+            signInResolverFactories: {
+              ...oidcProxySignInResolvers,
+              ...commonSignInResolvers,
+            },
+          }),
+        });
+        logger.info('auth backend-module holos-proxy-provider initialized');
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/module.ts:59]:
```suggestion
        logger.info('auth backend-module oidc-proxy-provider initialized');
```

but ... maybe just skip this log entirely?

File: plugins/auth-backend-module-oidc-proxy-provider/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export { authModuleOidcProxyProvider as default } from './module';
+export { oidcProxySignInResolvers } from './resolvers';
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/index.ts:18]: you typically want to export the underlying authenticator too, that's [what others do](https://github.com/backstage/backstage/blob/732583700dd0809dae4614792bd67177d7c9246e/plugins/auth-backend-module-oauth2-proxy-provider/src/index.ts#L24) so that people can build their own implementations of your provider

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+): (token: string) => Promise<IDTokenInfo> {
+  let jwksUri: string = '';
+
+  return async function tokenValidator(token) {
+    // Perform discovery once to avoid two round trips on every authentication request.
+    if (jwksUri === '' && providedJwks === undefined) {
+      jwksUri = await discoverJwksUri(iss);
+    }
+
+    logger.info(`jwksUri: ${jwksUri}`);
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:37]: remove

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    // Verify the token was signed by the issuer.  Performs one round trip to
+    // the jwks uri every authentication request to fetch the current key set
+    // from the issuer.  May be optimized in the future, but this happens only
+    // at sign-in, not for every request to the backend.
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:41]: sure, but once for every user, and in an org with thousands of people it may start to get noticeable

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    //
+    // Refer to https://github.com/panva/jose/blob/v5.4.0/docs/functions/jwks_remote.createRemoteJWKSet.md#returns-1
```
Comment (freben) [inline plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:43]: I'm still not entirely clear on why you didn't just use `createRemoteJWKSet` - that's what we do (passing the key set as the second argument to `jwtVerify`), and it keeps itself in sync perfectly as far as we know in the background, refreshing itself from the remote when needed etc.

**Review by freben — COMMENTED** [review]
Thanks!

**Comment by jeffmccune** [comment]
I'm working on this again, please re-open.

File: docs/auth/index.md [hunk 16]
```diff
@@ -38,6 +38,7 @@ Backstage comes with many common authentication providers in the core library:
 - [Google IAP](google/gcp-iap-auth.md)
 - [Okta](okta/provider.md)
 - [OAuth 2 Custom Proxy](oauth2-proxy/provider.md)
+- [OIDC Proxy](oidc/oidc-proxy.md)
```
[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/index.md:41]: Done in 305475c1b374 `docs: rename oidc-proxy.md to oidc/provider.md`

File: docs/auth/oidc/oidc-proxy.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
+---
+id: oidc-proxy-auth
```
[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/oidc/oidc-proxy.md:2]: Fixed

File: docs/auth/oidc/oidc-proxy.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+Backstage allows offloading the responsibility of authenticating users to a service mesh. The `oidcProxy` provider is developed and intended for use with the [Holos](https://openinfrastructure.co/holos/) platform, which integrates Backstage with Istio [External Authorization](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/).
+
+The authentication and authorization process are generally useful, however, and are not specific to Holos or to Istio. Any HTTP proxy operating as an identity aware proxy (IAP) capable of providing an oidc id token in a requiest header is compatible with this `oidcProxy` provider. For example, [NGINX](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/), [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy), and Envoy [External Authorization](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter) are all capable of providing the id token in the http request.
+
+It is assumed an IAP is already serving traffic in front of a Backstage instance configured to serve the frontend app from the backend.
```
[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/oidc/oidc-proxy.md:15]: Replacing line 15 with "It is assumed an IAP is already handling traffic for Backstage."

File: docs/auth/oidc/oidc-proxy.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+```yaml
+auth:
+  providers:
+    oidc-proxy:
+      # issuer is required, must match the iss claim of the id token.
+      issuer: https://login.example.com
+      # audience is required, must match one of the aud values listed in the id token.
+      audience: 'https://backstage.example.com'
+      # oidcIdTokenHeader is optional and may be set to any value except `authorization`.
+      # The value of the header is the id token without a "Bearer " prefix.
```
[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/oidc/oidc-proxy.md:31]: No, what I was trying to convey is the id token must be the value without any prefixes/schemes.  There isn't any support for slicing off prefixes.  I'll clarify.

[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/oidc/oidc-proxy.md:31]: Changed to "The value of the header is the id token without any prefix or suffix."

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts (hunk 1 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:21]: Changed to "Configuration for the oidc identity aware proxy auth provider."

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts (hunk 2 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:26]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts (hunk 3 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:31]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts (hunk 4 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:36]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/config.d.ts (hunk 5 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/config.d.ts:24]: It's a reflection of [gcpIap?: {](https://github.com/jeffmccune/backstage/blob/305475c1b374cae3302fdfdad97c1be296e13af5/plugins/auth-backend-module-gcp-iap-provider/config.d.ts#L23-L24)

Let me know if you want something different, otherwise leaving this as-is.

File: plugins/auth-backend-module-oidc-proxy-provider/package.json (hunk 6 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/package.json:4]: Removed

File: plugins/auth-backend-module-oidc-proxy-provider/README.md (hunk 7 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/README.md:1]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/CHANGELOG.md
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/CHANGELOG.md:1]: Removed

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts (hunk 8 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:27]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts (hunk 9 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:32]: Renamed

File: plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts (hunk 10 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/authenticator.ts:42]: Fixed

File: plugins/auth-backend-module-oidc-proxy-provider/src/module.ts (hunk 11 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/module.ts:59]: Removed

File: plugins/auth-backend-module-oidc-proxy-provider/src/index.ts (hunk 12 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/index.ts:18]: Exported to align with the others.

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts (hunk 13 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:37]: Removed

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts (hunk 14 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:41]: This is a reflection of your comment in the gcp-iap provider, which behaves the same:  [// TODO(freben): Rate limit the public key reads. It may be sensible to](https://github.com/jeffmccune/backstage/blob/305475c1b374cae3302fdfdad97c1be296e13af5/plugins/auth-backend-module-gcp-iap-provider/src/helpers.ts#L28-L32)

Let me know if you'd like it changed, and what to, otherwise I'll leave it.

File: plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts (hunk 15 above)
Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:43]: This code here is a reflection of me taking the code from the gcp-iap provider [here](https://github.com/jeffmccune/backstage/blob/305475c1b374cae3302fdfdad97c1be296e13af5/plugins/auth-backend-module-gcp-iap-provider/src/helpers.ts#L32) which uses the cloud sdk.

The cloud SDK, is GCP specific, but it's still OIDC under the hood, so I dove into _their_ source code to figure out how they verified the id token.

I then re-implemented what Google did natively here, which is why we ended up with this method.

In summary, an attempt on my part to remain faithful to gcp-iap, but perhaps missed the mark as soon as I started doing it the Google way instead of the other ways.

Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:43]: Anyway, it works, so I'd rather not mess with it unless there's a compelling reason to.  Let me know.

Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:43]: Ah, I recall now.  The other reason is the identity provider Holos uses is ZITADEL.  ZITADEL sometimes returns an empty jwks, `[]`, which caused `createRemoteJWKSet` to fail without grace, which in turn caused Backstage to fail with a mysterious user-facing error.

So, this is handles that edge case.

Refer to https://github.com/zitadel/zitadel/issues/6949#issuecomment-2035030448

Comment (jeffmccune) [reply plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts:43]: I added a comment to clarify the intent and why createRemoteJWKSet is avoided.

**Comment by jeffmccune** [comment]
Continued in https://github.com/backstage/backstage/pull/25851

File: docs/auth/oidc/oidc-proxy.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+id: oidc-proxy-auth
+title: OIDC Proxy Provider
+sidebar_label: OIDC Proxy
+# prettier-ignore
+description: Using an OIDC ID token from a request header as an authentication provider in Backstage
+---
+
+This tutorial shows how to authenticate and sign in users who access Backstage with an oidc id token already present in the incoming http request from their browser.
+
+Backstage allows offloading the responsibility of authenticating users to a service mesh. The `oidcProxy` provider is developed and intended for use with the [Holos](https://openinfrastructure.co/holos/) platform, which integrates Backstage with Istio [External Authorization](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/).
```
[the remark this answers is not in the document]
Comment (jeffmccune) [reply docs/auth/oidc/oidc-proxy.md:11]: Fixed
